cmd/skopeo/sync.go

```diff
@@ -0,0 +1,509 @@
+	var dest string
+
+	if srcRef.Transport() == transports.Get("docker") {
+		dest = fmt.Sprintf("%s/%s", globalDest, srcRef.DockerReference())
+	} else {
+		// It's `dir` transport. In that case`globalDest` uses the `docker://` transport because
+		// we don't allow `dir` -> `dir` sync operations.
+		relPath := strings.TrimPrefix(srcRef.StringWithinTransport(), rootSrc.StringWithinTransport())
+		// we cannot use `filepath.Join` against a "vanilla" globalDest because it will change
+		// `docker://` into `docker:/` breaking the reference validations later on
+		dest = fmt.Sprintf(
+			"docker://%s",
+			filepath.Join(strings.TrimPrefix(globalDest, "docker://"), relPath))
+	}
+
+	if strings.HasPrefix(dest, "dir:") {
+		// the final directory holding the image must exist otherwise
+		// the directory ImageReference instance won't be created
+		tgtDir := filepath.Dir(strings.TrimPrefix(dest, "dir:"))
+
+		if _, err := os.Stat(tgtDir); err != nil {
+			if os.IsNotExist(err) {
+				if err := os.MkdirAll(tgtDir, 0755); err != nil {
+					return nil, fmt.Errorf("Error creating directory for image %s: %v",
+						tgtDir,
+						err)
+				}
+			} else {
+				return nil, fmt.Errorf("Error while checking existance of directory %s: %v",
+					tgtDir,
+					err)
+			}
+		}
+	}
+
+	return getImageReference(dest)
+}
+
+func getImageTags(ctx context.Context, sysCtx *types.SystemContext, imgRef types.ImageReference) (tags []string, retErr error) {
+	img, err := imgRef.NewImage(ctx, sysCtx)
+	if err != nil {
+		return tags, err
+	}
+	defer func() {
+		if err := img.Close(); err != nil {
+			retErr = errors.Wrapf(retErr, fmt.Sprintf("(could not close image: %v) ", err))
+		}
+	}()
+	if dockerImg, ok := img.(*docker.Image); ok {
+		logrus.WithFields(logrus.Fields{
+			"image": dockerImg.SourceRefFullName(),
+		}).Info("Getting tags")
+		tags, retErr = dockerImg.GetRepositoryTags(context.Background())
+		if retErr != nil {
+			// some registries may decide to block the "list all tags" endpoint
+			// gracefully allow the inspect to continue in this case. Currently
+			// the IBM Bluemix container registry has this restriction.
+			if !strings.Contains(retErr.Error(), "401") {
+				return tags, fmt.Errorf("Error determining repository tags: %v", retErr)
+			}
+			logrus.Warn("Registry disallows tag list retrieval; skipping")
+		}
+	}
+
+	return
+}
+
+// Return true if the image had a tag specified, false otherwise
+func isTagSpecified(image string) (bool, error) {
+	if strings.HasSuffix(image, ":latest") {
+		return true, nil
+	}
+
+	// Normalize the image name, this will automatically add
+	// the `latest` tag when no tag has been specified
+	normName, err := reference.ParseNormalizedNamed(image)
+	if err != nil {
+		return true, err
+	}
+
+	// if the tag is `latest` -> the tag has been automatically added -> no tag
+	// was specified by the user
+	return !strings.HasSuffix(reference.TagNameOnly(normName).String(), ":latest"), nil
```

Comment:
(I’m sorry about not looking at this for so long. I do hope to make a proper review soonish.)

One general comment: podman and buildah have similarly built fair amount of code around manipulating the string forms of image names, and have been rather struggling to correctly handle the various corner cases and ambiguous forms of input.

I _very strongly_ want to avoid introducing such fragility to skopeo.

As a very local case, this should not be using `.String()` and then `HasSuffix` to parse it, but an interface conformance test against `reference.NamedTagged`. [Or maybe something else entirely, to also handle digested references.] 

_Most_ of the string parsing and formatting in here looks, at a first glance, like code that should not be using strings (though I haven’t read enough of the PR to be know that it is actually avoidable, to be fair.)

Reply:
Thanks, it turned out that most the raw strings manipulation could be avoided by rewriting some parts of the code and making use of [containers/image](https://github.com/containers/image) libraries.